cmapi/cmapi_server/managers/certificate.py

```diff
@@ -0,0 +1,106 @@
+        key = rsa.generate_private_key(
+            public_exponent=65537,
+            key_size=2048,
+            backend=default_backend()
+        )
+
+        with open(KEY_FILENAME, "wb") as f:
+            f.write(key.private_bytes(
+                encoding=serialization.Encoding.PEM,
+                format=serialization.PrivateFormat.TraditionalOpenSSL,
+                encryption_algorithm=serialization.NoEncryption()),
+            )
+
+        subject = issuer = x509.Name([
+            x509.NameAttribute(NameOID.COUNTRY_NAME, 'US'),
+            x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, 'California'),
+            x509.NameAttribute(NameOID.LOCALITY_NAME, 'Redwood City'),
+            x509.NameAttribute(NameOID.ORGANIZATION_NAME, 'MariaDB'),
+            x509.NameAttribute(NameOID.COMMON_NAME, 'mariadb.com'),
+        ])
+
+        basic_contraints = x509.BasicConstraints(ca=True, path_length=0)
+
+        cert = x509.CertificateBuilder(
+        ).subject_name(
+            subject
+        ).issuer_name(
+            issuer
+        ).public_key(
+            key.public_key()
+        ).serial_number(
+            x509.random_serial_number()
+        ).not_valid_before(
+            datetime.now(timezone.utc)
+        ).not_valid_after(
+            datetime.now(timezone.utc) + timedelta(days=CERT_DAYS)
+        ).add_extension(
+            basic_contraints,
+            False
+        ).add_extension(
+            x509.SubjectAlternativeName([x509.DNSName('localhost')]),
+            critical=False
+        ).sign(key, hashes.SHA256(), default_backend())
+
+        with open(CERT_FILENAME, 'wb') as f:
+            f.write(cert.public_bytes(serialization.Encoding.PEM))
+        logging.info('Created self signed sertificate for CMAPI API access.')
+    
+    @staticmethod
+    def create_self_signed_certificate_if_not_exist() -> None:
+        """Create self-signed certificate if not exist."""
+        if not os.path.exists(CERT_FILENAME):
+            CertificateManager.create_self_signed_certificate()
+
+    @staticmethod
+    def days_before_expire() -> int:
+        """Calculates how many days before expiration left.
+
+        Creates self-signed cetificate if certificate doesn't exist.
+
+        :return: days left
+        :rtype: int
+        """
+        CertificateManager.create_self_signed_certificate_if_not_exist()
+        with open(CERT_FILENAME, 'rb') as cert_file:
+            cert_data = cert_file.read()
+        cert = x509.load_pem_x509_certificate(cert_data)
+        days_before_expire = (cert.not_valid_after - datetime.now()).days
+        return days_before_expire
+
+    @staticmethod
+    def renew_certificate() -> None:
+        """Creates new self signed certificate.
+        
+        Creates self-signed cetificate if certificate doesn't exist or
+        expires in a 1 day or less.
+        """
+        CertificateManager.create_self_signed_certificate_if_not_exist()
```

Comment:
I might be too strict but "create_self_signed_certificate_if_not_exist" shouldn't be a part of "renew" and you should explicitly call it before "renew". Otherwise there is an implicit call that a reader must find on his own.

Reply:
Now I see that you decided to use this method as a guard. I don't like this approach personally but I am OK with it, so plz disregard my prev comment on this line.

Reply:
Ok, got it, but why you don't like such an approach?
Pros and cons?

Reply:
An implicit action that is a side effect mostly unexpected. One has to read the code of the functions to realize this.

Reply:
Done.